feat(sharding): add shard info args to requestCommunityInfo calls

jrainville opened the pull request:
Fixes #12015

Hooks the shard info args to the `requestCommunityInfo` functions.

Also, fixes the deep link code. It was calling it before the main module was ready and also it was using the old URL format.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | a60da019 | [#1](https://ci.status.im/job/status-desktop/job/prs/job/macos/job/aarch64/job/package/job/PR-12675/1/) | 2023-11-09 16:04:28 | ~7 min | `macos/aarch64` | [:apple:`dmg`](https://status-im-desktop-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Desktop-231109-155711-a60da0-pr12675-aarch64.dmg) |

igor-sirotin commented on src/app/modules/startup/module.nim:
```diff
@@ -326,10 +326,11 @@ method emitObtainingPasswordSuccess*[T](self: Module[T], password: string) =
 
 method finishAppLoading*[T](self: Module[T]) =
   self.delegate.finishAppLoading()
+  self.delegate.appReady()
```
hmmm, looks like we could use `finishAppLoading`, why have 2 signals?

jrainville replied:
It felt easier, but also it's clearer that we call this only when the app is ready, while `finishAppLoading` is to load the app.

igor-sirotin commented on src/app/modules/main/chat_section/chat_content/messages/controller.nim:
```diff
@@ -240,8 +243,10 @@ proc getCommunityDetails*(self: Controller): CommunityDto =
 
-proc requestCommunityInfo*(self: Controller, communityId: string, useDataabse: bool, requiredTimeSinceLastRequest: Duration) =
-  self.communityService.requestCommunityInfo(communityId, false, useDataabse, requiredTimeSinceLastRequest)
+proc requestCommunityInfo*(self: Controller, communityId: string, useDataabse: bool,
```
```suggestion
proc requestCommunityInfo*(self: Controller, communityId: string, useDatabase: bool,
```

igor-sirotin commented on src/app/modules/shared_models/link_preview_model.nim:
```diff
@@ -302,6 +302,12 @@ QtObject:
       if communityId != "":
         result.incl(communityId)
 
+  proc getCommunityLinks*(self: Model): Table[string, string] =
```
I think `getCommunityIds` is not needed anymore?

igor-sirotin reviewed (changes requested):
Nice job, many fixes and cleanups!

There's also `ImportCommunityPopup` that uses `requestCommunityInfo` function and we might have `shard` info there if it's a URL input:
https://github.com/status-im/status-desktop/blob/a60da0194ea78fc852bbc6e3b2035fc2b7856b39/ui/imports/shared/popups/ImportCommunityPopup.qml#L70

jrainville commented on src/app_service/service/shared_urls/dto/url_data.nim:
```diff
@@ -56,12 +61,26 @@ proc toContactUrlDataDto*(jsonObj: JsonNode): ContactUrlDataDto =
+  if (jsonObj.getProp("shard", shardObj)):
+    result.shard = shardObj.toShardUrlDataDto()
+  else:
+    result.shard = ShardUrlDataDto(
```
I tried, but Nim doesn't let me. We can't pass `nil` to something that has a type. That might just be missing knowledge from my part?

igor-sirotin replied:
Hmm, I don't get it. It can be a pointer to `Shard`, nope?

jrainville replied:
You mean doing this?
```nim
else:
    result.shard = Shard()
```
If so, I prefer setting the values to -1, as that way we can easily check the values before sending to status-go. I think they would default to 0 otherwise and those can exist I think

igor-sirotin replied:
I mean that if we change the definition of `Shard` to be a `ref object`:
```nim
type Shard* =  ref object
```

... or make the particular field to be `ref`, but this will require more changes, I'd go with first way.
```nim
  shard*: ref Shard
```

Then u can with no problems set `result.shard = nil`. Tested, compiles 😄

jrainville replied:
Updated @igor-sirotin you can review again

jrainville commented:
> There's also ImportCommunityPopup that uses requestCommunityInfo function and we might have shard info there if it's a URL input:

I didn't touch that one because the UI doesn't support putting the shard yet as far as I can tell. I wondered if maybe we should just drop the Import popup feature. Right now it only accepts an ID, but it would need new fields and at that point, is that worth it?

jrainville commented:
@igor-sirotin thanks for the review. I think I addresses all your comments.

Some of the ones I didn't were basically because of a limitation in Nim er maybe in my understanding of Nim. I can't just pass Nil as an Object as far as I can see.
Anyway, I use an async task to call the bakend function, so I wouldn't be able to see Shard to the thread as far as I know.

igor-sirotin commented:
> I didn't touch that one because the UI doesn't support putting the shard yet as far as I can tell. I wondered if maybe we should just drop the Import popup feature. Right now it only accepts an ID, but it would need new fields and at that point, is that worth it?

It actually supports pasting a shared url. That's why I noted about it 😄 

https://github.com/status-im/status-desktop/blob/b0e77c0966e7a442fe8507c31a04bc5e23939d66/ui/imports/shared/popups/ImportCommunityPopup.qml#L41

<img width="1312" alt="image" src="https://github.com/status-im/status-desktop/assets/25482501/b809a08f-7134-4b5f-96f5-de790685c7e2">

jrainville commented:
Good catch @igor-sirotin . I thought that popup only supported the ID, but I updated to get the shard data from the URL.

Also, do you know if we still use the InvitationBubbleView component? That is the one that is used when we use the Share Community feature, but I think you removed that feature now? So maybe we can get rid of it? It also calls `requestCommunityInfo`,  but just has access to the communityId it seems.

igor-sirotin commented:
> Also, do you know if we still use the InvitationBubbleView component? That is the one that is used when we use the Share Community feature, but I think you removed that feature now? So maybe we can get rid of it? It also calls `requestCommunityInfo`, but just has access to the communityId it seems.

@jrainville I think it's this one. We still have it, but should drop: https://github.com/status-im/status-desktop/issues/12570
So I don't think it makes sense add shards suppport to it.

osmaczko reviewed (commented):
Bearing https://github.com/status-im/status-desktop/issues/12585 in mind, perhaps we should reconsider whether we really need these changes. Maybe we don't need information about shards in links at all.

jrainville commented:
> Bearing #12585 in mind, perhaps we should reconsider whether we really need these changes. Maybe we don't need information about shards in links at all.

I had in mind that https://github.com/status-im/status-desktop/issues/12585 would work more as a back up than a default. So if possible, both would use the `requestCommunityInfo` function, but in the case of missing shard info, we'd recover the shard and then recursively call `requestCommunityInfo` again with it.

The reason to use the shard directly in the `requestCommunityInfo` function is just reliability and performance. It's gonna be way better UX if the link contains the shard info and we use it straight away, then request the shard info from waku, then requesting the community description. That latter one has two points of failures instead of one.

One edge case is if the community shard changes, then the old link would no longer be right.
What we can do in that case is that if we can't find the community description with the shard info, we can try again with the default shard. If we still find nothing, too bad, we give up, if we do find something and it's a new shard, we try that shard instead.

osmaczko commented:
> > Bearing #12585 in mind, perhaps we should reconsider whether we really need these changes. Maybe we don't need information about shards in links at all.
> 
> I had in mind that #12585 would work more as a back up than a default. So if possible, both would use the `requestCommunityInfo` function, but in the case of missing shard info, we'd recover the shard and then recursively call `requestCommunityInfo` again with it.
> 
> The reason to use the shard directly in the `requestCommunityInfo` function is just reliability and performance. It's gonna be way better UX if the link contains the shard info and we use it straight away, then request the shard info from waku, then requesting the community description. That latter one has two points of failures instead of one.
> 
> One edge case is if the community shard changes, then the old link would no longer be right. What we can do in that case is that if we can't find the community description with the shard info, we can try again with the default shard. If we still find nothing, too bad, we give up, if we do find something and it's a new shard, we try that shard instead.

I somewhat agree about reliability and performance. On the other hand, there wouldn't be much difference, and I doubt it would be visible to the end user (for instance, fetching a link in 2 seconds vs. 4 seconds).

More concerning, in my view, are the obsolete links to the community. For example, if we share a link to the Status Community on Twitter, ideally, we would need to update the link once we change the shard. Yes, we have a fallback mechanism, but this will introduce other issues and complexity.

In my opinion, we should stick to the simplest and most predictable version. The fewer arguments and less convoluted logic, the better. We could also prevent some areas from propagating shard information (like links, contracts, and possibly others in the future).

It's just my viewpoint, not something I'm insisting on.

igor-sirotin commented on src/app/modules/main/profile_section/profile/module.nim:
```diff
@@ -186,15 +187,12 @@ method updateProfileShowcasePreferences(self: Module, preferences: ProfileShowca
-      self.controller.requestCommunityInfo(communityEntry.communityId)
-      profileCommunityItems.add(initProfileShowcaseCommunityLoadingItem(
-        communityEntry.communityId, communityEntry.showcaseVisibility, communityEntry.order))
+      warn "Unknown community added to our own profile showcase" , communityId = community.id
```
could it be that we received a profile showcase update with new community, but didn't yet receive that we joined the new joined community? I mean receiving from a paired device.

igor-sirotin commented on src/app_service/common/types.nim:
```diff
@@ -76,10 +76,11 @@ type
     Completed
 
-type Shard* = object
+type Shard* = ref object
```
❤️

igor-sirotin commented on src/app_service/service/community/dto/community.nim:
```diff
@@ -467,6 +467,8 @@ proc toCommunityDto*(jsonObj: JsonNode): CommunityDto =
     discard shardObj.getProp("index", shard.index)
     result.shard = shard
+  else:
+    result.shard = nil
```
Which we already have! 🎉 
https://github.com/status-im/status-desktop/blob/cab78d9f08887f71777342447dd22b57ae3fe7f3/src/app_service/service/shared_urls/dto/url_data.nim#L60-L63

maybe even internnally check if `jsonObj.kind == JsonNull` and return `nil`.

jrainville commented on src/app/modules/main/shared_urls/module.nim:
```diff
@@ -53,8 +53,12 @@ method parseSharedUrl*(self: Module, url: string): UrlDataDto =
 method parseCommunitySharedUrl*(self: Module, url: string): string =
-  let communityData = self.controller.parseCommunitySharedUrl(url)
-  return $communityData
+  let parsedUrl = self.controller.parseSharedUrl(url)
+  let shard = parsedUrl.shard
+  var communityJson = parsedUrl.community.toJsonNode()
+  communityJson["shardCluster"] = %shard.cluster
+  communityJson["shardIndex"] = %shard.index
+  return $communityJson
```
Yeah, I agree that it would be better if it was part of the original object in status-go. So yeah, I had to hack this up to get the shard info to the QML from the URL. I created this issue to track it in status-go: https://github.com/status-im/status-go/issues/4318

I'll try to merge it in Nim for now

jrainville commented:
@igor-sirotin updated. Thanks again for the comments.

jrainville commented on src/backend/communities.nim:
```diff
@@ -433,12 +433,24 @@ proc collectCommunityMetrics*(communityId: string, metricsType: int, intervals:
-    "tryDatabase": tryDatabase,
-    "waitForResponse": true
-  }])
+proc requestCommunityInfo*(communityId: string, tryDatabase: bool, shardCluster: int, shardIndex: int): RpcResponse[JsonNode] {.raises: [Exception].} =
```
I tried. First, since I use an async_task, I still needed to recreate the `Shard` object since I can only pass strings and ints to the threads. Also it crashed so I said fuck it lol

igor-sirotin reviewed (approved):
LGTM now! Thanks for all the changes and patience 😄
